File: i2f-jdk/i2f-swl/src/main/java/i2f/swl/core/key/impl/SwlMemKeyManager.java
package i2f.swl.core.key.impl;

import i2f.jce.std.encrypt.asymmetric.key.AsymKeyPair;
import i2f.swl.core.key.SwlKeyManager;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicReference;

/**
 * @author Ice2Faith
 * @date 2024/8/8 21:15
 * @desc
 */
public class SwlMemKeyManager implements SwlKeyManager {
    protected ConcurrentHashMap<String, AsymKeyPair> selfCache = new ConcurrentHashMap<>(1024);
    protected ConcurrentHashMap<String, String> otherCache = new ConcurrentHashMap<>(1024);
    protected AtomicReference<String> selfDefaultCache = new AtomicReference<>();
    protected AtomicReference<String> otherDefaultCache = new AtomicReference<>();


    @Override
    public AsymKeyPair getSelfKeyPair() {
        String selfAsymSign = getDefaultSelfAsymSign();
        return getSelfKeyPair(selfAsymSign);
    }

    @Override
    public String getDefaultSelfAsymSign() {
        return selfDefaultCache.get();
    }

    @Override
    public void setDefaultSelfAsymSign(String selfAsymSign) {
        selfDefaultCache.set(selfAsymSign);
    }

    @Override
    public AsymKeyPair getSelfKeyPair(String selfAsymSign) {
        if (selfAsymSign == null) {
            return null;
        }
        AsymKeyPair ret = selfCache.get(selfAsymSign);
        return ret == null ? null : ret.copy();
    }

    @Override
    public void setSelfKeyPair(String selfAsymSign, AsymKeyPair keyPair) {
        if (selfAsymSign == null || keyPair == null) {
            return;
        }
        selfCache.put(selfAsymSign, keyPair.copy());
    }

    @Override
    public String getOtherPublicKey() {
        String otherAsymSign = getDefaultOtherAsymSign();
        return getOtherPublicKey(otherAsymSign);
    }

    @Override
    public String getDefaultOtherAsymSign() {
        return otherDefaultCache.get();
    }

    @Override
    public void setDefaultOtherAsymSign(String otherAsymSign) {
        otherDefaultCache.set(otherAsymSign);
    }

    @Override
    public String getOtherPublicKey(String otherAsymSign) {
        return otherCache.get(otherAsymSign);
    }

    @Override
    public void setOtherPublicKey(String otherAsymSign, String publicKey) {
        otherCache.put(otherAsymSign, publicKey);
    }
}

File: i2f-jdk/i2f-swl/web/i2f-swl/src/i2f-turbo-web/i2f-swl/core/core/key/impl/SwlMemKeyManager.js
import SwlKeyManager from "../SwlKeyManager";
import SwlBase64Obfuscator from "../../../impl/SwlBase64Obfuscator";
import AsymKeyPair from "../../../../../i2f-core/crypto/asymmetric/AsymKeyPair";

/**
 * @return {SwlMemKeyManager}
 * @implements {SwlKeyManager}
 * @constructor {SwlMemKeyManager}
 */
function SwlMemKeyManager() {

    this.selfCache=new Map()

    this.otherCache=new Map()

    this.selfDefaultCache=null

    this.otherDefaultCache=null
}

// 继承
SwlMemKeyManager.prototype = Object.create(SwlKeyManager.prototype)
SwlMemKeyManager.prototype.constructor = SwlMemKeyManager


/**
 * @return {AsymKeyPair}
 */
SwlMemKeyManager.prototype.getDefaultSelfKeyPair = function () {
    let selfAsymSign = this.getDefaultSelfAsymSign();
    return this.getSelfKeyPair(selfAsymSign);
}

/**
 * @return {String}
 */
SwlMemKeyManager.prototype.getDefaultSelfAsymSign = function () {
    return this.selfDefaultCache;
}

/**
 *
 * @param selfAsymSign {String}
 * @return {void}
 */
SwlMemKeyManager.prototype.setDefaultSelfAsymSign = function (selfAsymSign) {
    this.selfDefaultCache=selfAsymSign;
}

/**
 *
 * @param selfAsymSign {String}
 * @return {AsymKeyPair}
 */
SwlMemKeyManager.prototype.getSelfKeyPair = function (selfAsymSign) {
    if (!selfAsymSign) {
        return null;
    }
    let ret = this.selfCache[selfAsymSign];
    return ret?ret.copy():null
}

/**
 *
 * @param selfAsymSign {String}
 * @param keyPair {AsymKeyPair}
 * @return {void}
 */
SwlMemKeyManager.prototype.setSelfKeyPair = function (selfAsymSign, keyPair) {
    if (!selfAsymSign || !keyPair) {
        return;
    }
    this.selfCache[selfAsymSign]=keyPair.copy();
}

/**
 * @return {String}
 */
SwlMemKeyManager.prototype.getDefaultOtherPublicKey = function () {
    let otherAsymSign = this.getDefaultOtherAsymSign();
    return this.getOtherPublicKey(otherAsymSign);
}

/**
 * @return {String}
 */
SwlMemKeyManager.prototype.getDefaultOtherAsymSign = function () {
    return  this.otherDefaultCache;
}

/**
 *
 * @param otherAsymSign {String}
 * @return {void}
 */
SwlMemKeyManager.prototype.setDefaultOtherAsymSign = function (otherAsymSign) {
    this.otherDefaultCache=otherAsymSign;
}

/**
 *
 * @param otherAsymSign {String}
 * @return {String}
 */
SwlMemKeyManager.prototype.getOtherPublicKey = function (otherAsymSign) {
    if (!otherAsymSign) {
        return null;
    }
    return this.otherCache[otherAsymSign];

}

/**
 *
 * @param otherAsymSign {String}
 * @param publicKey {String}
 * @return {void}
 */
SwlMemKeyManager.prototype.setOtherPublicKey = function (otherAsymSign, publicKey) {
    if (!otherAsymSign || !publicKey) {
        return;
    }
    this.otherCache[otherAsymSign]=publicKey;
}

export default SwlMemKeyManager
